templates/boltzmann/middleware/honeycomb.ts

```diff
@@ -9,77 +10,133 @@ import beeline from 'honeycomb-beeline'
 import isDev from 'are-we-dev'
 void `{% endif %}`;
 
+let otelTracer: OtelTracer | null = null;
+
+function getOtelTracer() {
+  if (!otelTracer) {
+    otelTracer = otelTrace.getTracer('boltzmann', '1.0.0')
+  }
+  return otelTracer
+}
+
 function trace ({
   headerSources = ['x-honeycomb-trace', 'x-request-id'],
 } = {}) {
   if (!process.env.HONEYCOMB_WRITEKEY) {
     return (next: Handler) => (context: Context) => next(context)
   }
 
-  const schema = require('honeycomb-beeline/lib/schema')
-  const tracker = require('honeycomb-beeline/lib/async_tracker')
-
-  return function honeycombTrace (next: Handler) {
-    return (context: Context) => {
-      const traceContext = _getTraceContext(context)
-      const trace = beeline.startTrace({
-        [schema.EVENT_TYPE]: 'boltzmann',
-        [schema.PACKAGE_VERSION]: '1.0.0',
-        [schema.TRACE_SPAN_NAME]: `${context.method} ${context.url.pathname}${context.url.search}`,
-        [schema.TRACE_ID_SOURCE]: traceContext.source,
-        'request.host': context.host,
-        'request.original_url': context.url.href,
-        'request.remote_addr': context.remote,
-        'request.method': context.method,
-        'request.scheme': context.url.protocol,
-        'request.path': context.url.pathname,
-        'request.query': context.url.search
-      },
-      traceContext.traceId,
-      traceContext.parentSpanId,
-      traceContext.dataset)
-
-      if (isDev()) {
-        context._honeycombTrace = trace
-      }
-
-      if (traceContext.customContext) {
-        beeline.addContext(traceContext.customContext)
-      }
+  if (process.env.HONEYCOMB_API_HOST) {
+    return function honeycombOtelTrace (next: Handler) {
+      return (context: Context) => {
+        const activeContext = otelContext.active()
+        const tracer = getOtelTracer()
+
+        const rootSpan = tracer.startSpan(`${context.method} ${context.url.pathname}${context.url.search}`, {
+          attributes: {
+            'request.host': context.host,
+            'request.original_url': context.url.href,
+            'request.remote_addr': context.remote,
+            'request.method': context.method,
+            'request.scheme': context.url.protocol,
+            'request.path': context.url.pathname,
+            'request.query': context.url.search
+          }
+        }, activeContext)
+
+        otelTrace.setSpan(activeContext, rootSpan)
+
+        // do not do as I do,
+        onHeaders(context._response, function () {
+          rootSpan.setAttribute('response.status_code', String(context._response.statusCode))
+          if (context.handler) {
+            const handler = context.handler;
+            if (handler.route) {
```

Comment:
These ifs are driven by typescript's type checker, which was beyond cranky at the thought that these things could potentially be undefined. It's weird to me that my editor didn't complain about the analogous beelines code, but hey.

Reply:
I've handled this elsewhere using typecasts – e.g. `rootSpan.setAttribute('request.route', <string>handler.route)`

Reply:
I'll give it a try and see how it feels.